clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -134,14 +134,25 @@ public String toString() {
     }
 
     /**
-     * Expire the batch that is ready but is sitting in accumulator for more than requestTimeout due to metadata being unavailable.
-     * We need to explicitly check if the record is full or linger time is met because the accumulator's partition may not be ready
-     * if the leader is unavailable.
+     * A batch whose metadata is not available should be expired if one of the following is true:
+     * <ol>
+     *     <li> the batch is not in retry AND request timeout has eplapsed after it is ready. (We need to see if a batch is
+     *     ready by explicitly checking if the record is full or linger time is met because the accumulator's partition
+     *     may not be ready if the leader is unavailable.)
+     *     <li> the batch is in retry AND request timeout has elapsed after the backoff period ended.
+     * </ol>
      */
-    public boolean maybeExpire(int requestTimeout, long now, long lingerMs) {
+    public boolean maybeExpire(int requestTimeoutMs, long retryBackoffMs, long now, long lingerMs) {
         boolean expire = false;
-        if ((this.records.isFull() && requestTimeout < (now - this.lastAppendTime)) || requestTimeout < (now - (this.lastAttemptMs + lingerMs))) {
+
+        if (!this.inRetry() && this.records.isFull() && requestTimeoutMs < (now - this.lastAppendTime))
+            expire = true;
+        else if (!this.inRetry() && requestTimeoutMs < (now - (this.createdMs + lingerMs)))
             expire = true;
+        else if (this.inRetry() && requestTimeoutMs < (now - (this.lastAttemptMs + retryBackoffMs)))
+            expire = true;
+
+        if (expire) {
             this.records.close();
             this.done(-1L, Record.NO_TIMESTAMP, new TimeoutException("Batch Expired"));
```

Comment:
Maybe you can change this line as per the following PR since you're changing this method anyway:

https://github.com/apache/kafka/pull/695/files#diff-d40f2124414371535f2f1e47089eb71dR146